README.md
remoteStorage Starter Kit
=========================

This starter-kit is the quickest way to get started with remoteStorage app
development. It gives you a local remoteStorage server and a Hello World app
as your starting point.

In order to run it, you need
[node.js](http://nodejs.org/download/) installed.

First, save this starter-kit into a folder on your computer, by downloading
and unpacking https://github.com/remotestorage/starter-kit/archive/master.zip
or by cloning this repository using git.

Then, open a command-line terminal, and type:

    cd starter-kit
    npm install
    
to install the dependencies (where `starter-kit` is the folder you unpacked from
the download or with git).

This starter-kit contains three things, baked into one:

  * a remoteStorage server, which you can use as `me@localhost:8001`
  * a launch screen, which you can open at `http://localhost:8002/`
  * some sample app, in the `apps/` folder (where the launch screen preloads from).

To launch all three, you only need to type one command:

    node starter-kit


### Next steps

* Visit `http://localhost:8002/` and click on one of the icons
* Open the web console (Ctrl-Shift-K in FF, Ctrl-Shift-I in Chrome, F12 in IE)
* There are now various server origins interacting:
  * the API of your remoteStorage server is on http://localhost:8000/storage/me/
  * the html page you used with the links to the apps runs on http://localhost:8002/
  * the actual apps run on http://localhost:8003/, http://localhost:8004/, etc.
* see how the AJAX requests go to the remoteStorage API on port 8000, which is different
    from the origin of the app you are viewing. It is a cross-origin AJAX request.
* This is of course useless if it's on the same host, but you can see how this is
    a powerful architecture change if not only the port but also the domain name
    differs between the app and the storage: netizens can host their own data on
    their own server, instead of on the application provider's server. See
    https://unhosted.org/ and http://nobackend.org/ for more info about this revolution.
* Inspect the application code in `apps/hello-world/index.html` using your favorite editor
* Inspect the `remoteStorage.notes` code in `apps/hello-world/notes.js`
* Read the docs on http://remotestorage.io/integrate/
* Hack! :) Copy and rename the `apps/hello-world/` folder to e.g. `apps/my-first-unhosted-app/`
* The starter-kit will detect each folder you create under `apps/` at startup, and add
    it to the launch screen on `http://localhost:8002/` (make sure to restart `node starter-kit`
    each time you edit your app on disk).
* Post your reactions and questions on 
    http://community.remotestorage.io/category/getting-started
* TIP: if you had trouble installing, this [screencast](//www.youtube.com/embed/eGNJRyb5iJs) may help.

### Web authoring

The starter-kit also supports cloning apps from elsewhere, as long as those apps are hosted with CORS,
and have an `assets` array in their manifest. At the bottom of the portal page, you will see a big blue
"INSTALL NEW UNHOSTED WEB APPS" button. Currently, only the 'my favorite drinks' app is available there
for cloning, as a demo, but you could add your own app in two ways:

* by sending a pull request to https://github.com/unhosted/store/blob/master/defaultApps.json
* by specifying a different app channel URL than the default https://apps.unhosted.org/defaultApps.json

Just make sure the manifest contains an 'assets' array, otherwise it will not be clonable. Since you have
full control over your remoteStorage data, you can also edit apps after you cloned them. To see this in
action, type:

````js
remoteStorage.www.storeFile('8003', 'text/html', 'index.html',
    '<html><h2>editing my app!</h2></html>');
````

in the console while on the portal page, and then visit http://localhost:8003/ to see the result. Web authoring
is a new optional feature of the remoteStorage spec, which we have been talking about for several years now (also
in the context of [Read-Write Web](http://www.w3.org/community/rww/), and whose development was funded by NLNet.
The new spec describing it (version 04) will come out in December 2014, and it has been added as an *optional*
feature, meaning not all server implementations and storage providers may support it. Whether or not the currently
connected account supports web authoring can be detected through `remoteStorage.remote.properties`, which is
supported by remotestorage.js 0.11 and up.


CHANGELOG.md
# 0.2.2

* support state parameter in OAuth dance
* update to remotestorage-server 0.5.11
* add ludbud example

# 0.2.1

* update to remotestorage-server 0.5.6

# 0.2.0

* second version

# 0.1.0

* first version


starter-kit.js
var fs = require('fs'),
  http = require('http'),
  url = require('url');

var config = {
  defaultUserName: 'me',
  storageRoot: '/storage/',
  webAuthoringPath: '/public/www/',
  host: 'localhost',
  storagePort: 8000,
  mainPort: 8001,
  portalPort: 8002,
  firstAppPort: 8003,
  firstCloningPort: 1024,
  lastCloningPort: 1034,
  apps: {}
};

var server;

function contentTypeFromFilename(fileName) {
  if (fileName.substr(-5) === '.html') {
    return new Buffer('text/html', 'utf-8');
  } else if (fileName.substr(-4) === '.css') {
    return new Buffer('text/css', 'utf-8');
  } else if (fileName.substr(-3) === '.js') {
    return new Buffer('application/javascript', 'utf-8');
  } else {
    return new Buffer('text/plain', 'utf-8');
  }
}

function loadFiles(dir, port, basedir, callback) {
  var list = fs.readdirSync(dir),
    assets = [], numToDo = list.length;
  if (!basedir) {
    basedir = dir;
  }
  function doneOne() {
    numToDo--;
    if (numToDo ===0) {
      callback(assets);
    }
  }
  list.forEach(function(fileName) {
    var filePath = dir + '/' + fileName;
    var pathOnStorage;
    var stat = fs.statSync(filePath);
    if (stat && stat.isDirectory()) {
      loadFiles(filePath, port, basedir, function(subAssets) {
        for(var i =0; i<subAssets.length; i++) {
          assets.push(filePath + '/' + subAssets[i]);
        }
        doneOne();
      });
    } else {
     pathOnStorage = config.webAuthoringPath + port.toString() + filePath.substring(basedir.length),
      server.backdoorSet(config.defaultUserName,
          pathOnStorage,
          fs.readFileSync(filePath),
          contentTypeFromFilename(fileName),
          function(err, revision) {
            assets.push(filePath);
            doneOne();
          });
    }
  });
}

function websiteServer(filePath, port, callback) {
  if(filePath) {
    loadFiles(filePath, port, undefined, callback);
  }//else it will be empty until an app uses the www
  //module to store a cloned app for that port.

  return function (req, res) {
    req.url = config.storageRoot + config.defaultUserName + config.webAuthoringPath + port.toString() + req.url;
    if (req.url.substr(-1) === '/') {
      req.url += 'index.html';
    }
    return serveStorage(req, res);
  };
}

function setApps(listing) {
  server.backdoorSet(config.defaultUserName,
    '/apps/channel-url',
    new Buffer('https://apps.unhosted.org/defaultApps.json', 'utf-8'),
    new Buffer('application/json', 'utf-8'),
    function(err, revision) {
      console.log('channel url stored stored', '/apps/channel-url');
    }
  );
  for(var i=0; i<listing.length; i++) {
    var listener = websiteServer('./apps/'+listing[i], config.firstAppPort+i, function(bindName, bindPort) {
      return function(assets) {
        server.backdoorSet(config.defaultUserName,
          '/apps/' + bindName,
          new Buffer(JSON.stringify({
            name: bindName,
            href: 'http://localhost:' + bindPort + '/',
            img: 'http://localhost:' + bindPort + '/icon_x128.png',
            assets: assets,
            cloned: true
          }), 'utf-8'),
          new Buffer('application/json', 'utf-8'),
          function(err, revision) {
            console.log('manifest stored', '/apps/' + bindName + '.manifest');
          }
        );
      };
    }(listing[i], config.firstAppPort+i));
    http.createServer(listener).listen(config.firstAppPort+i);
    config.apps['http://'+config.host+':'+(config.firstAppPort+i)+'/'] = listing[i];
  }
}

function prepareCloningPorts() {
  for(var i=config.firstCloningPort; i <= config.lastCloningPort; i++) {
    var listener = websiteServer(undefined, i, function() {});
    http.createServer(listener).listen(i);
  }
}

function setPortal() {
  var listener = websiteServer('./portal', config.portalPort, function(portalAssets) {
  });
  http.createServer(listener).listen(config.portalPort);
}

var kv = (function() {
  var store = {};
  return {
    set: function(k, v, cb) { store[k] = v; cb(); },
    get: function(k, cb) { cb(null, store[k]); }
  };
 })();

function serveMain(req, res) {
  var urlObj = url.parse(req.url, true), userAddress, userName;
  if(urlObj.pathname == '/') {
    server.portal(req, res);
  } else if(urlObj.pathname == '/.well-known/webfinger') {
    server.webfinger(req, res);
  } else if(urlObj.pathname.substring(0, '/auth/'.length) == '/auth/') {
    server.oauth(req, res);
  } else {
    res.writeHead(404);
    res.end();
  }
}

function serveStorage(req, res) {
  var urlObj = url.parse(req.url, true), userAddress, userName;
  console.log(urlObj.pathname);
  if(urlObj.pathname.substring(0, '/storage/'.length) == '/storage/') {
    server.storage(req, res);
  } else {
    res.writeHead(404);
    res.end();
  }
}

function launch() {
  fs.readdir('./apps/', function(err, listing) {
    if(err) {
      console.log('make sure ./apps/ exists');
    } else {
      console.log('Loading files from the "apps/" folder into the "apps" and "www" modules:');
      server = require('./localhost-server').createInstance(kv, config);
      setApps(listing);
      prepareCloningPorts();
      setPortal();
      http.createServer(serveMain).listen(config.mainPort);
      http.createServer(serveStorage).listen(config.storagePort);
      console.log('\nSee http://' + config.host + ':' + config.portalPort + '/\n'
          + '\nor visit a http-hosted, remoteStorage.js-based app and connect with me@localhost:'+ config.mainPort
          + '\n(special backdoor, does not work from https-hosted apps!)\n'
          + '\nNB: User data is only stored in memory, not saved to disk.');
    }
  });
}

//...
launch();
